rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md

```diff
@@ -0,0 +1,336 @@
+supported the type of behavior users may be accustomed to from those
+other tools.
+
+While users could technically just increase Vector's batch sizes and
+timeouts to approach the desired behavior, the implementation is not
+designed for it and would cause significant problems:
+
+- End-to-end acknowledgements are not issued until the batch is
+  finished, which will cause backpressure to sources and possibly
+  timeouts with retransmissions.
+- The batches are stored in memory which could cause out of memory
+  crashes.
+
+## Proposal
+
+Vector will use the AWS S3 multipart upload API as the primary upload
+process. This allows the sink to persist data to S3 sooner, to
+acknowledge data within Vector earlier, and to reduce the memory
+required for buffering the upload batches.
+
+The AWS S3 multipart upload API uses the following process:
+
+1. The upload is created with the `CreateMultipartUpload` action. This
+   action returns a unique identifier for the upload which is used for
+   all following actions.
+2. Individual parts are uploaded with the `UploadPart` action. These
+   uploads contain a sequence number that identifies the order they
+   will be assembled into the final object, and may be uploaded in any
+   order.
+3. The final object is assembled with the `CompleteMultipartUpload`
+   action.
+
+There are two major issues that this API presents that force some of
+the implementation decisions below:
+
+1. Vector may need to continue a multipart upload process across sink
+   restarts (due to reloads, restarts, or crashes). Since the uploads
+   require the presence of a unique identifier for both the multipart
+   upload itself as well as each of the uploaded parts, Vector will
+   need to track these identifiers with the partition that initiated
+   the upload.
+2. The upload parts (except for the final part) have a _minimum_ size
+   limit of 5MB. Since the sink may need to be restarted before a
+   batch reaches this minimum size, vector will need to complete the
+   multipart object when shutting down.
+
+In the case of an interruption in the shutdown process or a crash, the
+destination bucket will be left with incomplete multipart uploads and
+parts. To recover from this, Vector will optionally scan for such
+parts at startup and assemble them to their original object
+names. That process works as follows:
+
+1. The existing incomplete uploads are listed with the
+   `ListMultipartUploads` action.
+2. Included in the listing are the object names the uploads would be
+   written to. The `key_prefix` in the configuration is translated
+   into a pattern, and these object names are filtered to match only
+   those that could have come from the current configuration.
+3. For each resulting multipart upload, the parts that compose it are
+   listed with one or more `ListParts` actions.
+4. The upload assembled to the final object with the
+   `CompleteMultipartUpload` action as above.
```

Comment:
Just verifying, is this operation sensitive to race conditions? What if multiple Vector instances all do this at once?

Reply:
If multiple Vector instances are configured to write to the same bucket with the same `key_prefix`, then yes one could reassemble uploads another is still working on. I can't think of a good way to avoid this other than either persisting the objects the current system is working on (which I understood to be undesirable, and wouldn't handle the case of a crashed node that never comes back) or manually configuring the recovery to only happen on one instance. Possibly a timeout could work (ie only reassemble uploads older than some threshold), but that too has issues if there is a slow writer.

Reply:
I think it is fine if another Vector instance finalizes the upload. My question, similar to Ben's, is what would happen if two Vector instances both tried to finalize the upload? My guess is that one of them would fail because the multi part upload was completed by another instance. I think this is fine. We can just output a warning in that case.

Another case to consider is Vector A starting up and finding a partial upload started by Vector B (that is still running). Assuming that Vector A finalizes B's batch, how would B be affected?

One more edge case, what if Vector crashes and then is started with a new prefix?

I also think we'll want to clearly document this lest people suffer data loss because Vector crashed, but they never restarted it.

Reply:
Agree, I'm ok if one of the instances fails assuming the operation succeeds on another. We could have thoughtful retries too. That is, if an operation fails, check if the operation is still needed before retrying. Finally, this is a consideration for central state management and coordination (AKA the enterprise product).

Reply:
If two Vector instances try to finalize the same object, I think only one would succeed. It should be very difficult to test this, though, as it is a race condition requiring the second request to come in before AWS S3 finishes its operation.

The second edge case is the more troubling one. In that case, Vector A could trash B's upload. Hmm, although in that case the data is still there on S3, B would just need to start a new upload and continue. I'll add that to the process description.

If Vector crashes and is restarted with a new prefix, and no other instance shares the same prefix, then the upload would never be reassembled.

Reply:
We should peruse the docs and verify this won't have conflicts, because we raise the chances significantly in scenarios where many Vector instances boot at once.

Reply:
I tested the handling of adding to a deleted multipart upload, and it produces a HTTP 404 as expected, which is something we can catch as described in the latest update. With that process, I don't see a race/conflict possibility that wouldn't have to be handled by S3 itself.

Reply:
Great, this should be resolved then.